// @graphql-mesh/soap/index.mjs
import { PredefinedProxyOptions } from '@graphql-mesh/store';
import { SOAPLoader, createExecutorFromSchemaAST } from '@omnigraph/soap';
import { readFileOrUrl } from '@graphql-mesh/utils';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import { parse, buildASTSchema } from 'graphql';

class SoapHandler {
    constructor({ config, store, baseDir, importFn, logger }) {
        this.config = config;
        this.soapSDLProxy = store.proxy('schemaWithAnnotations.graphql', PredefinedProxyOptions.StringWithoutValidation);
        this.baseDir = baseDir;
        this.importFn = importFn;
        this.logger = logger;
    }
    async getMeshSource({ fetchFn }) {
        const soapSDL = await this.soapSDLProxy.getWithSet(async () => {
            const soapLoader = new SOAPLoader({
                fetch: fetchFn,
            });
            const location = this.config.wsdl;
            const wsdl = await readFileOrUrl(location, {
                allowUnknownExtensions: true,
                cwd: this.baseDir,
                fetch: fetchFn,
                importFn: this.importFn,
                logger: this.logger,
            });
            const object = await soapLoader.loadWSDL(wsdl);
            soapLoader.loadedLocations.set(location, object);
            const schema = soapLoader.buildSchema();
            return printSchemaWithDirectives(schema);
        });
        const schemaAST = parse(soapSDL);
        const executor = createExecutorFromSchemaAST(schemaAST, fetchFn);
        const schema = buildASTSchema(schemaAST);
        return {
            schema,
            executor,
        };
    }
}

export default SoapHandler;
